mantle/cmd/kola/devshell.go

```diff
@@ -144,195 +105,321 @@ func runDevShellSSH(ctx context.Context, builder *platform.QemuBuilder, conf *co
+				sc.controlChan <- sshStart //signal that SSH is ready
+				statusMsg = "QEMU guest is ready for SSH"
+			} else {
+				ready = false
+				sc.controlChan <- sshNotReady // signal that SSH should be terminated
 
-	// Ignore other status messages, and just print errors for now
-	go func() {
-		for {
-			select {
-			case <-serialChan:
-			case <-readychan:
-			case err := <-errchan:
-				fmt.Fprintf(os.Stderr, "errchan: %v", err)
+				switch state {
+				case guestStateInShutdown:
+					statusMsg = "QEMU guest is shutting down"
+				case guestStateHalted:
+					statusMsg = "QEMU guest is halted"
+					inst.Kill()
+				case guestStateInReboot:
+					statusMsg = "QEMU guest initiated reboot"
+				case guestStateOpenSshStopped:
+					statusMsg = "QEMU openssh is not listening"
+				case guestStateSshDisconnected:
+					statusMsg = "SSH Client disconnected"
+				case guestStateStopHang:
+					statusMsg = "QEMU guest waiting for a systemd unit to stop"
+				case guestStateBooting:
+					statusMsg = "QEMU guest is booting"
+				}
 			}
-		}
-	}()
+			displayStatusMsg("EVENT", statusMsg)
 
-	var ip string
-	err = util.Retry(6, 5*time.Second, func() error {
-		var err error
-		ip, err = inst.SSHAddress()
-		if err != nil {
-			return err
+		// monitor the SSH connection
+		case err := <-sc.errChan:
+			if err == nil {
+				sc.kill()
+				displayStatusMsg("SESSION", "Clean exit from SSH, terminating instance")
+				return nil
+			}
+			displayStatusMsg("SSH Terminated", fmt.Sprintf("Error: %v", err))
 		}
-		return nil
-	})
-	if err != nil {
-		return errors.Wrapf(err, "awaiting ssh address")
 	}
+}
 
-	sshConfigPath := filepath.Join(tmpd, "ssh-config")
-	sshConfig, err := os.OpenFile(sshConfigPath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
-	if err != nil {
-		return errors.Wrapf(err, "creating ssh config")
-	}
-	defer sshConfig.Close()
-	sshBuf := bufio.NewWriter(sshConfig)
+// guestStates are uses to report guests events internal to the host.
+const (
+	// guestStateOpenSshStarted indicates that the guest has started SSH and
+	// is ready for an SSH connection.
+	guestStateOpenSshStarted = iota
+	// guestStateOpenSshStopped indicates that the SSH server has transitioned
+	// from running to stopped.
+	guestStateOpenSshStopped
+	// guestStateSshDisconnected indicates the SSH client has disconnected
+	guestStateSshDisconnected
+	// guestStateInShutdown indicates that the guest is shutdown
+	guestStateInShutdown
+	// guestStateInReboot indicates that the guest has started a reboot
+	guestStateInReboot
+	// guestStateHalted indicates that the guest has halted or shutdown
+	guestStateHalted
+	// guestStateBooting indicates that the instance is in early boot
+	guestStateBooting
+	// guestStateStopHang indicates a job is being slow to shutdown
+	guestStateStopHang
+)
 
-	_, err = fmt.Fprintf(sshBuf, "Host %s\n", devshellHostname)
-	if err != nil {
-		return err
+// checkWriteState checks magical (shutter) strings for state. This is imprecise at best,
```

Comment:
I think what we were doing previously in parsing the systemd journal is at least cleaner.  And because the systemd journal is structured data we can even e.g. check for `openssh.service` in the systemd unit field rather than parsing the "OpenSSH server" string from the unit description (which could easily change). But it's tricky because we kind of do need to parse the console until we get the journal up.

Reply:
> And because the systemd journal is structured data we can even e.g. check for openssh.service in the systemd unit field rather than parsing the "OpenSSH server" string from the unit description 

True, however, the previous code didn't handle the structured data in a structure way. 

That said, the last logic for evaluating the JSON and checking for events. In my testing, relying on the journal is wonky since we have to intercept the shutdown of the journal in time; we still need to parse the serial log for magic strings.